Emit error for duplicate type parameters

```scala
class Test[A, A]: { x: A }
//│ Defined class Test

t = Test { x = 42 }
//│ t: Test['A | 'A0 .. 42 & 'A | 'A & 'A0, 'A | 'A0 .. 42 & 'A | 'A & 'A0] with {x: 42}

t : Test[bool, int]
//│ res: Test[bool | int .. nothing, bool | int .. nothing]

t : Test[int, bool]
//│ ╔══[ERROR] Type mismatch in type ascription:
//│ ║  l.10: 	t : Test[int, bool]
//│ ║        	^
//│ ╟── expression of type `42` does not match type `bool`
//│ ║  l.4: 	t = Test { x = 42 }
//│ ║       	               ^^
//│ ╟── but it flows into reference with expected type `Test[int, bool]`
//│ ║  l.10: 	t : Test[int, bool]
//│ ║        	^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.10: 	t : Test[int, bool]
//│ ║        	              ^^^^
//│ ╟── from record type:
//│ ║  l.1: 	class Test[A, A]: { x: A }
//│ ╙──     	                  ^^^^^^^^
//│ res: Test[bool | int .. nothing, bool | int .. nothing]
```
Currently the last one takes precedence, and the output contains duplicate type variables. (It breaks surprisingly few things besides ascription.)

--- a/shared/src/main/scala/mlscript/Typer.scala
+++ b/shared/src/main/scala/mlscript/Typer.scala
@@ -252,6 +252,11 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
       }
       if (!n.name.head.isUpper) err(
         msg"Type names must start with a capital letter", n.toLoc)
+      td.tparams.groupBy(_.name).foreach { case s -> tps if tps.size > 1 => err(
+          msg"Multiple declarations of type parameter ${s} in ${td.kind.str} definition" -> td.toLoc
+            :: tps.map(tp => msg"Declared at" -> tp.toLoc))
+        case _ =>
+      }
       val dummyTargs = td.tparams.map(p =>
         freshVar(originProv(p.toLoc, s"${td.kind.str} type parameter"), S(p.name))(ctx.lvl + 1))
       val td1 = TypeDef(td.kind, td.nme, td.tparams, dummyTargs, td.body, td.mthDecls, td.mthDefs, 
@@ -429,6 +434,11 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
         thisCtx += "this" -> subst(thisCtx.thisType(tn.name), rigid)
         (td.mthDecls ++ td.mthDefs).foreach { case md @ MethodDef(rec, prt, nme, tparams, rhs) =>
           implicit val prov: TypeProvenance = tp(md.toLoc, rhs.fold(_ => "method definition", _ => "method declaration"))
+          tparams.groupBy(_.name).foreach { case s -> tps if tps.size > 1 => err(
+              msg"Multiple declarations of type parameter ${s} in ${prov.desc}" -> md.toLoc
+                :: tps.map(tp => msg"Declared at" -> tp.toLoc))
+            case _ =>
+          }
           val dummyTargs2 = tparams.map(p =>
             TraitTag(Var(p.name))(originProv(p.toLoc, "method type parameter")))
           val targsMap2 = tparams.map(_.name).zip(dummyTargs2).toMap
@@ -557,11 +567,18 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
           else ComposedType(true, rec(lhs), rec(rhs)) _
         )(tp(ty.toLoc, "union type"))
       case Neg(t) => NegType(rec(t))(tp(ty.toLoc, "type negation"))
-      case Record(fs) => RecordType.mk(fs.map { nt =>
+      case Record(fs) => 
+        val prov = tp(ty.toLoc, "record type")
+        fs.groupMap(_._1.name)(_._1).foreach { case s -> fieldNames if fieldNames.size > 1 => err(
+            msg"Multiple declarations of field name ${s} in ${prov.desc}" -> ty.toLoc
+              :: fieldNames.map(tp => msg"Declared at" -> tp.toLoc))(raise, prov)
+          case _ =>
+        }
+        RecordType.mk(fs.map { nt =>
           if (nt._1.name.head.isUpper)
-            err(msg"Field identifiers must start with a small letter", nt._1.toLoc)(raise, noProv)
+            err(msg"Field identifiers must start with a small letter", nt._1.toLoc)(raise, prov)
           nt._1 -> rec(nt._2)
-        })(tp(ty.toLoc, "record type"))
+        })(prov)
       case Function(lhs, rhs) => FunctionType(rec(lhs), rec(rhs))(tp(ty.toLoc, "function type"))
       case WithExtension(b, r) => WithType(rec(b),
         RecordType(r.fields.mapValues(rec))(tp(r.toLoc, "extension record")))(tp(ty.toLoc, "extension type"))
@@ -750,11 +767,17 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
       case App(App(Var("&"), lhs), rhs) =>
         typeTerm(lhs) & (typeTerm(rhs), prov)
       case Rcd(fs) => // TODO rm: no longer used?
+        val prov = tp(term.toLoc, "record literal")
+        fs.groupMap(_._1.name)(_._1).foreach { case s -> fieldNames if fieldNames.size > 1 => err(
+            msg"Multiple declarations of field name ${s} in ${prov.desc}" -> term.toLoc
+              :: fieldNames.map(tp => msg"Declared at" -> tp.toLoc))(raise, prov)
+          case _ =>
+        }
         RecordType.mk(fs.map { case (n, t) => 
-            if (n.name.head.isUpper)
-              err(msg"Field identifiers must start with a small letter", term.toLoc)(raise, noProv)
-            (n, typeTerm(t))
-          })(tp(term.toLoc, "record literal"))
+          if (n.name.head.isUpper)
+            err(msg"Field identifiers must start with a small letter", term.toLoc)(raise, prov)
+          (n, typeTerm(t))
+        })(prov)
       case tup: Tup =>
         typeTerms(tup :: Nil, false, Nil)
       case Bra(false, trm: Blk) => typeTerm(trm)

--- a/shared/src/test/diff/mlscript/BadMethods.mls
+++ b/shared/src/test/diff/mlscript/BadMethods.mls
@@ -486,3 +486,65 @@ Test6C
 //│ ║  l.437: 	class Test6C: Test6A[int] & Test6B
 //│ ╙──       	      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error
+
+
+
+
+:e
+class Dup[A, A]: { x: A }
+    method MthDup[B, C, // random comment
+        B]: (A -> B) -> B
+    method MthDup f = f this.x
+//│ ╔══[ERROR] Multiple declarations of type parameter A in class definition
+//│ ║  l.494: 	class Dup[A, A]: { x: A }
+//│ ║         	      ^^^^^^^^^^^^^^^^^^^
+//│ ╟── Declared at
+//│ ║  l.494: 	class Dup[A, A]: { x: A }
+//│ ║         	          ^
+//│ ╟── Declared at
+//│ ║  l.494: 	class Dup[A, A]: { x: A }
+//│ ╙──       	             ^
+//│ ╔══[ERROR] Multiple declarations of type parameter B in method declaration
+//│ ║  l.495: 	    method MthDup[B, C, // random comment
+//│ ║         	           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+//│ ║  l.496: 	        B]: (A -> B) -> B
+//│ ║         	^^^^^^^^^^^^^^^^^^^^^^^^^
+//│ ╟── Declared at
+//│ ║  l.495: 	    method MthDup[B, C, // random comment
+//│ ║         	                  ^
+//│ ╟── Declared at
+//│ ║  l.496: 	        B]: (A -> B) -> B
+//│ ╙──       	        ^
+//│ Defined class Dup
+//│ Declared Dup.MthDup: Dup['A | 'A0 .. 'A & 'A0, 'A | 'A0 .. 'A & 'A0] -> ('A0 -> 'B) -> 'B
+//│ Defined Dup.MthDup: Dup['A | 'A0 .. 'A & 'A0, 'A | 'A0 .. 'A & 'A0] -> ('A0 -> 'a) -> 'a
+
+t = Dup { x = 42 }
+//│ t: Dup['A | 'A0 .. 42 & 'A | 'A & 'A0, 'A | 'A0 .. 42 & 'A | 'A & 'A0] with {x: 42}
+
+t : Dup[bool, int]
+//│ res: Dup[bool | int .. nothing, bool | int .. nothing]
+
+t : Dup[int, bool]
+//│ ╔══[ERROR] Type mismatch in type ascription:
+//│ ║  l.528: 	t : Dup[int, bool]
+//│ ║         	^
+//│ ╟── expression of type `42` does not match type `bool`
+//│ ║  l.522: 	t = Dup { x = 42 }
+//│ ║         	              ^^
+//│ ╟── but it flows into reference with expected type `Dup[int, bool]`
+//│ ║  l.528: 	t : Dup[int, bool]
+//│ ║         	^
+//│ ╟── Note: constraint arises from type reference:
+//│ ║  l.528: 	t : Dup[int, bool]
+//│ ║         	             ^^^^
+//│ ╟── from record type:
+//│ ║  l.494: 	class Dup[A, A]: { x: A }
+//│ ╙──       	                 ^^^^^^^^
+//│ res: Dup[bool | int .. nothing, bool | int .. nothing]
+
+// SOF... (other than id, log, discard)
+// t.MthDup (fun x -> mul 2 x)
+
+t.MthDup id
+//│ res: 42

--- a/shared/src/test/diff/mlscript/BadMethods.mls
+++ b/shared/src/test/diff/mlscript/BadMethods.mls
@@ -543,7 +543,7 @@ t : Dup[int, bool]
 //│ ╙──       	                 ^^^^^^^^
 //│ res: Dup[bool | int .. nothing, bool | int .. nothing]
 
-// SOF... (other than id, log, discard)
+// SOF... (other than 'A -> T, anything -> T)
 // t.MthDup (fun x -> mul 2 x)
 
 t.MthDup id

--- a/shared/src/test/diff/mlscript/BadTypes.mls
+++ b/shared/src/test/diff/mlscript/BadTypes.mls
@@ -30,9 +30,43 @@ type Hello = int
 //│ res: error
 
 
-// TODO forbid
-aa = {a = 1; a = "oops" }
+:e
+aa = { a = 1; a = "oops" }
+//│ ╔══[ERROR] Multiple declarations of field name a in record literal
+//│ ║  l.34: 	aa = { a = 1; a = "oops" }
+//│ ║        	     ^^^^^^^^^^^^^^^^^^^^^
+//│ ╟── Declared at
+//│ ║  l.34: 	aa = { a = 1; a = "oops" }
+//│ ║        	       ^
+//│ ╟── Declared at
+//│ ║  l.34: 	aa = { a = 1; a = "oops" }
+//│ ╙──      	              ^
 //│ aa: {a: 1, a: "oops"}
+
+:e
+aa: { a: int; a: string }
+//│ ╔══[ERROR] Multiple declarations of field name a in record type
+//│ ║  l.47: 	aa: { a: int; a: string }
+//│ ║        	    ^^^^^^^^^^^^^^^^^^^^^
+//│ ╟── Declared at
+//│ ║  l.47: 	aa: { a: int; a: string }
+//│ ║        	      ^
+//│ ╟── Declared at
+//│ ║  l.47: 	aa: { a: int; a: string }
+//│ ╙──      	              ^
+//│ ╔══[ERROR] Type mismatch in type ascription:
+//│ ║  l.47: 	aa: { a: int; a: string }
+//│ ║        	^^
+//│ ╟── expression of type `1` does not match type `string`
+//│ ║  l.34: 	aa = { a = 1; a = "oops" }
+//│ ║        	           ^
+//│ ╟── but it flows into reference with expected type `{a: int, a: string}`
+//│ ║  l.47: 	aa: { a: int; a: string }
+//│ ║        	^^
+//│ ╟── Note: constraint arises from type reference:
+//│ ║  l.47: 	aa: { a: int; a: string }
+//│ ╙──      	                 ^^^^^^
+//│ res: {a: int, a: string}
 
 aa.a
 //│ res: 1

